src/domain/datasets/domain/src/repos/dataset_dependency_repository.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::pin::Pin;
+
+use internal_error::InternalError;
+use opendatafabric::DatasetID;
+use thiserror::Error;
+use tokio_stream::Stream;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait DatasetDependencyRepository: Send + Sync {
+    async fn stores_any_dependencies(&self) -> Result<bool, InternalError>;
+
+    fn list_all_dependencies(&self) -> DatasetDependenciesIDStream;
+
+    async fn add_upstream_dependencies(
+        &self,
+        downstream_dataset_id: &DatasetID,
+        new_upstream_dataset_ids: &[&DatasetID],
+    ) -> Result<(), AddDependenciesError>;
+
+    async fn remove_upstream_dependencies(
+        &self,
+        downstream_dataset_id: &DatasetID,
+        obsolete_upstream_dataset_ids: &[&DatasetID],
+    ) -> Result<(), RemoveDependenciesError>;
+
+    async fn remove_all_dependencies_of(&self, dataset_id: &DatasetID)
+        -> Result<(), InternalError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct DatasetDependencies {
+    pub downstream_dataset_id: DatasetID,
+    pub upstream_dataset_ids: Vec<DatasetID>,
```

Comment:
Why don't we stream just the ID pairs (as they are stored in the DB) and let the caller aggregate however they want? Seems like an unnecessary complication of a (what should be low-level) interface.

Reply:
We already had this modeling before. I don't see a motivation to reduce the level of abstraction.

Reply:
Inserts are looking quite nicely currently.